src/Particle/SoaDistanceTableABOMP.h

```diff
@@ -0,0 +1,426 @@
+                                                        first, last);
+        }
+    }
+  }
+
+  inline void mw_evaluate(const RefVector<DistanceTableData>& dt_list, const RefVector<ParticleSet>& p_list)
+  {
+    ScopedTimer eval(&eval_timer_);
+    mw_evaluate_fuse_transfer(dt_list, p_list);
+  }
+
+  inline void mw_evaluate_transfer_inplace(const RefVector<DistanceTableData>& dt_list, const RefVector<ParticleSet>& p_list)
+  {
+    const size_t nw = dt_list.size();
+    size_t count_targets = 0;
+    for (ParticleSet& p: p_list)
+      count_targets += p.getTotalNum();
+    const size_t total_targets = count_targets;
+
+    constexpr size_t realtype_size = sizeof(RealType);
+    constexpr size_t int_size = sizeof(int);
+    constexpr size_t ptr_size = sizeof(RealType*);
+    offload_input.resize(total_targets * D * realtype_size + total_targets * int_size + (nw + total_targets) * ptr_size);
+    auto target_positions = reinterpret_cast<RealType*>(offload_input.data());
+    auto walker_id_ptr = reinterpret_cast<int*>(offload_input.data() + total_targets * D * realtype_size);
+    auto source_ptrs = reinterpret_cast<RealType**>(offload_input.data() + total_targets * D * realtype_size + total_targets * int_size);
+    auto output_ptrs = reinterpret_cast<RealType**>(offload_input.data() + total_targets * D * realtype_size + total_targets * int_size + nw * ptr_size);
+
+    const int N_sources_padded = getAlignedSize<T>(N_sources);
+    offload_output.resize(total_targets * N_sources_padded * (D + 1));
+
+    count_targets = 0;
+    for (size_t iw = 0; iw < nw; iw++)
+    {
+      auto& dt = static_cast<SoaDistanceTableABOMP&>(dt_list[iw].get());
+      ParticleSet& pset(p_list[iw]);
+
+      assert(N_sources == dt.N_sources);
+
+      auto& RSoA_OMP = static_cast<const RealSpacePositionsOMP&>(dt.Origin->getCoordinates());
+      source_ptrs[iw] = const_cast<RealType*>(RSoA_OMP.getDevicePtr());
+
+      for (size_t iat = 0; iat < pset.getTotalNum(); ++iat, ++count_targets)
+      {
+        for (size_t idim = 0; idim < D; idim++)
+          target_positions[count_targets * D + idim] = pset.R[iat][idim];
+
+        walker_id_ptr[count_targets]   = iw;
+        output_ptrs[count_targets] = dt.r_dr_device_ptr_ + iat * N_sources_padded * (D + 1);
+      }
+    }
+
+    const int N_sources_local  = N_sources;
+    const int ChunkSizePerTeam = 256;
+    const int NumTeams         = (N_sources + ChunkSizePerTeam - 1) / ChunkSizePerTeam;
+
+    auto* input_ptr = offload_input.data();
+
+    {
+      ScopedTimer offload(&offload_timer_);
+      #pragma omp target teams distribute collapse(2) num_teams(total_targets*NumTeams) \
+        map(always, to: input_ptr[:offload_input.size()]) \
+        nowait depend(out: total_targets)
+      for (int iat = 0; iat < total_targets; ++iat)
+        for (int team_id = 0; team_id < NumTeams; team_id++)
+        {
+          auto* target_pos_ptr = reinterpret_cast<RealType*>(input_ptr);
```

Comment:
This a painful optimization I have to take. It put all the device pointers together and use a single transfer. This is done locally in the function and this can be allowed inside the OMP class.

Reply:
Add a comment along the lines "// This is horrible but allows all the device pointers to be copied in a single transfer".

Reply:
Added comment.